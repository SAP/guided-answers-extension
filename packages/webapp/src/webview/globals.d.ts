interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
