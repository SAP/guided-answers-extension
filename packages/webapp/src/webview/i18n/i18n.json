{
    "GUIDED_ANSWER_HEADER": "Guide",
    "SEARCH_GUIDED_ANSWERS": "Search Guided Answers",
    "ALL_ANSWERS": "Home",
    "STEP_BACK": "Step back",
    "RESTART": "Restart",
    "GUIDED_ANSWERS": "Guided Answers",
    "NO_ANSWERS_FOUND": "No answers found",
    "PLEASE_MODIFY_SEARCH": "Please modfiy search",
    "THIS_SOLVED_MY_ISSUE": "This solved my issue",
    "THIS_DID_NOT_SOLVED_MY_ISSUE": "This did not solve my issue",
    "THANK_YOU_TEXT": "We are glad to hear that your issue has been resolved and we hope you enjoy using the Guided Answers application.",
    "THANKS": "Thanks!",
    "HOME": "Home",
    "SEND": "Send",
    "CLOSE": "Close",
    "ISSUE_IS_NOT_RESOLVED": "Issue is not resolved",
    "WE_ARE_SORRY_TO_HEAR_THAT_YOUR_ISSUE_IS_NOT_YET_RESOLVED": "We are sorry to hear that your issue is not yet resolved.",
    "PLEASE_TELL_US_IF_THIS_ANSWER_WAS_HELPFUL": "Please tell us if this answer was helpful",
    "THERE_ARE_SEVERAL_OPTIONS_FOR_GETTING_FURTHER_ASSISTANCE": "There are several options for getting further assistance:",
    "SEARCH_FOR_ANOTHER_GUIDED_ANSWER": "Search for another Guided Answer",
    "START_AN_EXPERT_CHAT": "Start an Expert Chat",
    "SCHUEDULE_AN_EXPERT": "Schedule an Expert",
    "OPEN_AN_INCIDENT": "Open an Incident",
    "ASK_THE_SAP_COMMUNITY": "Ask the SAP Community",
    "FEEDBACK": "Feedback",
    "FEEDBACK_DIALOG_TITLE": "Is this content helpful?",
    "FEEDBACK_DIALOG_SUBTEXT": "If you have suggestions on how to improve this content we would love to hear them!",
    "FEEDBACK_DIALOG_DISCLAIMER": "Your feedback is anonymous, we do not collect any personal data.",
    "FEEDBACK_DIALOG_SUGGESTION": "Your suggestion",
    "FEEDBACK_SENT_DIALOG_TITLE": "Message sent",
    "FEEDBACK_SENT_DIALOG_SUBTEXT": "Thanks for the feedback!"
}
