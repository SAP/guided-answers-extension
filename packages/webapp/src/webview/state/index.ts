export { actions, store } from './store';
export { getInitialState } from './reducers';
