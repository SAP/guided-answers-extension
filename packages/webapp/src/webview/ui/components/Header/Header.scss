.guided-answer {
    &__header {
        display: flex;
        width: 100%;
        z-index: 10;
        background-color: var(--vscode-editor-background);
        position: sticky;
        top: 0;
        padding-bottom: 20px;
        &__sub {
            display: flex;
            align-items: center;
            width: 100%;
        }
        &__title {
            font-weight: 600;
            margin: 0;
            color: var(--vscode-settings-headerForeground);
            line-height: 29px;
        }
        &__subtitle {
            color: var(--vscode-editorCodeLens-foreground, #999);
            font-size: 2em;
        }
        &__navButtons {
            height: 25px;
            margin-top: 30px;
            color: var(--vscode-foreground);
            padding: 0 5px 0px 5px;
            text-align: start;
            cursor: pointer;
            background: none;
            border: 0;
            &__content {
                vertical-align: middle;
                display: inline-block;
                &__text {
                    margin-left: 3px;
                }
            }
            display: flex;
            align-items: center;
            svg {
                width: 16px;
                height: 16px;
            }
        }
        &__navButtons:hover {
            background: var(--vscode-menubar-selectionBackground);
            outline: 1px solid var(--vscode-contrastActiveBorder);
            border-radius: 3px;
            > span {
                text-decoration: none;
            }
        }
        &__navButtons:focus-visible {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
        }

        &__allAnswersButton {
            width: 25%;
            margin-right: 80px;
        }
        &__back-restart-buttons {
            display: flex;
            justify-content: space-between;
            width: 75%;
            max-width: 620px;
        }
        &__restart-feedback-buttons {
            display: flex;
        }
        &__divider {
            border-left: 1px solid var(--vscode-editorWidget-border);
            height: 16px;
            margin-top: 35px;
            margin-left: 10px;
            margin-right: 10px;
        }
        &__logoAndTitle {
            display: flex;
            align-items: center;
            margin-top: 30px;
            min-width: 280px;
        }
        &__searchField {
            display: flex;
            align-items: start;
            flex-direction: row;
            margin-top: 30px;
            width: 66%;
        }
    }
}

.home-icon {
    width: 62px;
}

@media only screen and (max-width: 625px) {
    .guided-answer {
        &__header {
            &__logoAndTitle {
                width: 50%;
            }

            &__sub {
                align-items: start;
                flex-direction: column;
            }

            &__searchField {
                margin-top: 20px;
                align-items: start;
            }
            &__navButtons {
                text-align: center;
                justify-content: center;
            }
            &__navButtons__content {
                margin-left: 0;
            }
            &__navButtons__content__text {
                display: none;
            }
            &__back-restart-buttons {
                justify-content: space-between;
                width: 100%;
            }
            &__allAnswersButton {
                margin-right: 48px;
                width: 28px;
                margin-left: 6px;
            }
        }
    }
    .home-icon {
        width: 28px;
        padding: 0 3px 2px 3px;
        border-radius: 3px;
        color: var(--vscode-menu-foreground, #cccccc);
    }
    .home-icon:hover {
        background: var(--vscode-menubar-selectionBackground);
    }
}
