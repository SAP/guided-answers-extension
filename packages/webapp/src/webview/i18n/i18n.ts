import { use } from 'i18next';
import { initReactI18next } from 'react-i18next';
import * as i18nEn from './i18n.json';

export { i18n } from 'i18next';

export function initI18n(): void {
    use(initReactI18next)
        .init({
            resources: {
                en: {
                    translation: i18nEn
                }
            },
            lng: 'en',
            fallbackLng: 'en',
            interpolation: {
                escapeValue: false
            }
        })
        .catch((error) => console.error(error));
}
