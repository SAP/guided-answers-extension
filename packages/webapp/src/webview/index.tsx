import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
