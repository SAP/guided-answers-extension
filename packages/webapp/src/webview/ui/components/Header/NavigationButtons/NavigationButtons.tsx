import React from 'react';
import { actions } from '../../../../state';
import { VscHome, VscRefresh, VscArrowLeft } from 'react-icons/vsc';
import i18next from 'i18next';
import { UIIcon } from '@sap-ux/ui-components';
import { UiIcons } from '../../UIComponentsLib/Icons';
import { focusOnElement } from '../../utils';

/**
 *
 * @returns A button component navigating to all answers.
 */
const AllAnswersButton = React.memo(function AllAnswersButton() {
    focusOnElement('.home-icon');

    return (
        <button
            className="guided-answer__header__navButtons home-icon"
            onClick={(): void => {
                actions.goToAllAnswers();
            }}
            title={i18next.t('ALL_ANSWERS')}>
            <VscHome className="guided-answer__header__navButtons__content" />{' '}
            <span className="guided-answer__header__navButtons__content__text text-underline">
                {i18next.t('ALL_ANSWERS')}
            </span>
        </button>
    );
});

/**
 *
 * @returns A button component for going back one step.
 */
export function BackButton() {
    return (
        <button
            className="guided-answer__header__navButtons"
            onClick={(): void => {
                actions.goToPreviousPage();
            }}
            title={i18next.t('STEP_BACK')}>
            <VscArrowLeft className="guided-answer__header__navButtons__content" />
            <span className="guided-answer__header__navButtons__content text-underline" style={{ marginLeft: '3px' }}>
                {i18next.t('STEP_BACK')}
            </span>
        </button>
    );
}

/**
 *
 * @returns A button component for going back to the first step.
 */
export function RestartButton() {
    return (
        <button
            className="guided-answer__header__navButtons"
            onClick={(): void => {
                actions.restartAnswer();
            }}
            title={i18next.t('RESTART')}>
            <VscRefresh className="guided-answer__header__navButtons__content" />
            <span className="guided-answer__header__navButtons__content text-underline" style={{ marginLeft: '3px' }}>
                {i18next.t('RESTART')}
            </span>
        </button>
    );
}

/**
 *
 * @returns A button component for feedback submission.
 */
export function GeneralFeedbackButton() {
    return (
        <button
            className="guided-answer__header__navButtons"
            onClick={(): void => {
                actions.feedbackStatus(true);
            }}
            title={i18next.t('FEEDBACK')}>
            <UIIcon
                className="guided-answer__header__navButtons__content"
                style={{ paddingTop: '2px' }}
                iconName={UiIcons.Chat}
            />
        </button>
    );
}

export { AllAnswersButton };
