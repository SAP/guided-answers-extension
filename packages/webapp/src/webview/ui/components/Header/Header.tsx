import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../../types';
import { AllAnswersButton, BackButton, GeneralFeedbackButton, RestartButton } from './NavigationButtons';
import { Logo } from './Logo';
import './Header.scss';
import { SearchField } from './SearchField';
import { FocusZone } from '@fluentui/react-focus';

/**
 * Renders and returns the header section.
 *
 * @param props Props for Header component
 * @param props.showSub Show/Hide subsection of the Header
 * @param props.showNavButons Show/Hide navigation buttons
 * @param props.showLogo Show/Hide SAP logo
 * @param props.showSearch Show/Hide Search field
 * @returns - react element for header section
 */
export function Header(props: {
    showSub: boolean;
    showNavButons: boolean;
    showLogo: boolean;
    showSearch: boolean;
}): ReactElement {
    const appState = useSelector<AppState, AppState>((state) => state);
    return (
        <div className="guided-answer__header" style={{ paddingBottom: props.showNavButons === true ? '0' : '20px' }}>
            {props.showSub === true ? (
                <>
                    <div className="guided-answer__header__sub">
                        {!!props.showLogo && <Logo />}
                        {!!props.showSearch && <SearchField />}
                    </div>{' '}
                </>
            ) : (
                <></>
            )}

            {props.showNavButons === true ? (
                <>
                    <FocusZone isCircularNavigation={true} className="guided-answer__header">
                        <div className="guided-answer__header__allAnswersButton">
                            <AllAnswersButton />
                        </div>

                        <div className="guided-answer__header__back-restart-buttons">
                            <BackButton />
                            <div className="guided-answer__header__restart-feedback-buttons">
                                {appState.activeGuidedAnswerNode.length > 1 && <RestartButton />}
                                {appState.activeGuidedAnswerNode.length > 1 && (
                                    <div className="guided-answer__header__divider"></div>
                                )}
                                <GeneralFeedbackButton />
                            </div>
                        </div>
                    </FocusZone>
                </>
            ) : (
                <></>
            )}
        </div>
    );
}
