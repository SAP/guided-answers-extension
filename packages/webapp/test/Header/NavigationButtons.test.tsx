import React from 'react';
import { shallow } from 'enzyme';
import { actions } from '../../src/webview/state';
import { AllAnswersButton, BackButton, RestartButton } from '../../src/webview/ui/components/Header/NavigationButtons';
import i18next from 'i18next';
import { initI18n } from '../../src/webview/i18n';

jest.mock('../../src/webview/state', () => {
    return {
        actions: {
            goToAllAnswers: jest.fn(),
            goToPreviousPage: jest.fn(),
            restartAnswer: jest.fn()
        }
    };
});
describe('<AllAnswersButton />', () => {
    let wrapper: any;
    initI18n();
    beforeEach(() => {
        wrapper = shallow(<AllAnswersButton />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a AllAnswersButton component', () => {
        //Test UI rendering
        expect(wrapper.find('.guided-answer__header__navButtons').length).toBe(1);
        expect(wrapper.find('.guided-answer__header__navButtons').hasClass('home-icon')).toBe(true);
        expect(wrapper.find('.guided-answer__header__navButtons').props().onClick).toBeDefined();
        expect(wrapper.find('VscHome').length).toBe(1);
        expect(wrapper.find('.guided-answer__header__navButtons__content__text').hasClass('text-underline')).toBe(true);
        expect(wrapper.find('.guided-answer__header__navButtons__content__text').text()).toBe(i18next.t('ALL_ANSWERS'));

        //Test click event
        wrapper.find('.guided-answer__header__navButtons').simulate('click');
        expect(actions.goToAllAnswers).toBeCalled();
    });
});

describe('<BackButton />', () => {
    let wrapper: any;
    initI18n();
    beforeEach(() => {
        wrapper = shallow(<BackButton />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a BackButton component', () => {
        //Test UI rendering
        expect(wrapper.find('.guided-answer__header__navButtons').length).toBe(1);
        expect(wrapper.find('.guided-answer__header__navButtons').props().onClick).toBeDefined();
        expect(wrapper.find('VscArrowLeft').length).toBe(1);
        expect(wrapper.find('.guided-answer__header__navButtons__content').at(1).hasClass('text-underline')).toBe(true);
        expect(wrapper.find('.guided-answer__header__navButtons__content').at(1).text()).toBe(i18next.t('STEP_BACK'));

        //Test click event
        wrapper.find('.guided-answer__header__navButtons').simulate('click');
        expect(actions.goToPreviousPage).toBeCalled();
    });
});

describe('<RestartButton />', () => {
    let wrapper: any;
    initI18n();
    beforeEach(() => {
        wrapper = shallow(<RestartButton />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a RestartButton component', () => {
        //Test UI rendering
        expect(wrapper.find('.guided-answer__header__navButtons').length).toBe(1);
        expect(wrapper.find('.guided-answer__header__navButtons').props().onClick).toBeDefined();
        expect(wrapper.find('VscRefresh').length).toBe(1);
        expect(wrapper.find('.guided-answer__header__navButtons__content').at(1).hasClass('text-underline')).toBe(true);
        expect(wrapper.find('.guided-answer__header__navButtons__content').at(1).text()).toBe(i18next.t('RESTART'));

        //Test click event
        wrapper.find('.guided-answer__header__navButtons').simulate('click');
        expect(actions.restartAnswer).toBeCalled();
    });
});
