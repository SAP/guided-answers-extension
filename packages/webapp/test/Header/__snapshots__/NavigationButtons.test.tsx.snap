// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackButton /> Should render a BackButton component 1`] = `
<div>
  <button
    class="guided-answer__header__navButtons"
    id="back-button"
    title="Step back"
  >
    <i
      aria-hidden="true"
      class="ts-icon root-105"
      data-icon-name="GoLeft"
    />
    <span
      class="guided-answer__header__navButtons__text"
    >
      Step back
    </span>
  </button>
</div>
`;

exports[`<HomeButton /> Should render a HomeButton component 1`] = `
<div>
  <button
    class="guided-answer__header__navButtons"
    id="home-button"
    title="Home"
  >
    <i
      aria-hidden="true"
      class="ts-icon root-105"
      data-icon-name="Home"
    />
    <span
      class="guided-answer__header__navButtons__text"
    >
      Home
    </span>
  </button>
</div>
`;

exports[`<RestartButton /> Should render a RestartButton component 1`] = `
<div>
  <button
    class="guided-answer__header__navButtons"
    id="restart-button"
    title="Restart"
  >
    <i
      aria-hidden="true"
      class="ts-icon root-105"
      data-icon-name="Refresh"
    />
    <span
      class="guided-answer__header__navButtons__text"
    >
      Restart
    </span>
  </button>
</div>
`;

exports[`<ShareButton /> Should render a ShareButton component 1`] = `
<div>
  <button
    class="ms-Button ms-Button--icon is-checked guided-answer__header__navButtons root-166"
    data-is-focusable="true"
    id="callout-test-id"
    title="Share this guide"
    type="button"
  >
    <span
      class="ms-Button-flexContainer flexContainer-167"
      data-automationid="splitbuttonprimary"
    >
      <i
        aria-hidden="true"
        class="ms-Icon root-105 ms-Button-icon icon-169"
        data-icon-name="Link"
      />
    </span>
  </button>
  <span
    class="ms-layer"
  />
</div>
`;
